doc/design/logging.md

```diff
@@ -0,0 +1,356 @@
+   implementation can also not make use of functionality in e.g. `iox::cxx` or
+   `iox::posix`
+
+3. To have a minimal runtime overhead, string formatting shall only happen when
+   the log level is above the output threshold. This lazy evaluation shall also
+   apply to other potentially expensive operations like the result of a function
+   call.
+
+4. Additionally, the minimal log level should be configurable by a compile time
+   switch and everything below this log level should be compiled to no-ops. With
+   this, developer can make excessive use of the most verbose log level and
+   disable it by default. When debugging, this can be turned on and help to find
+   the root of the problem.
+
+5. There shall also be the possibility to alter the behavior of the logger via
+   environment variables. This can be functionality like setting the log level or
+   deciding whether the timestamps or file names shall be part of a log message.
+
+6. Since iceoryx is often used as transport layer in a higher level framework, it
+   shall be possible to replace the default logger at runtime and delegate the
+   log messages to the logger of such a framework. To prevent misuse, this can
+   only be done up to a specific time. Afterwards an error message shall be
+   logged in the new and old logger and the error handler shall be called with a
+   moderate error level.
+
+7. The default logger can be replaced via the platform abstraction in order to
+   use the dedicated logger for log messages which are emitted before it could
+   be replaced at runtime, e.g. from global objects before `main`.
+
+8. Additionally, there shall be a compile time option to forward all log messages
+   independent of the log level threshold for cases where the logger framework
+   does the filtering itself.
+
+9. For tests the default logger shall be replaced to suppress the log messages in
+   passed tests but output them for failed tests. This keeps the noise low but
+   still provides useful information when needed. Additionally there shall be an
+   environment variable to circumvent the suppression when desired.
+
+### Solution
+
+#### Class diagram
+
+![logging class diagram](../website/images/logging_classes.svg)
+
+#### Macro with lazy evaluation
+
+The buildup of the log message is only done when the log level is above the output
+threshold. This is accomplished by a macro with an incomplete if-statement.
+
+```cpp
+#define LAZY() if (cond)
+
+LAZY() expensiveFunctionCall();
+```
+
+In the example above `expensiveFunctionCall` is only executed when `cond` is `true`.
+If `cond` is a compile time constant set to `false`, the whole statement is compiled
+to a no-op and optimized away.
+
+This is the log macro with lazy evaluation
+
+```cpp
+#define IOX_LOG_INTERNAL(file, line, function, level)          \
+    if ((level) <= iox::log::Logger::minimalLogLevel()         \
+        && (iox::log::Logger::ignoreLogLevel()                 \
```

Comment:
The `ignoreLogLevel` function would be deprecated if we do not allow to change to log level after `initLogger`, which I would prefer